src/queue.rs

```diff
@@ -0,0 +1,810 @@
+            );
+            false
+        } else if used_ring
+            .checked_add(used_ring_size as GuestUsize)
+            .map_or(true, |v| !mem.address_in_range(v))
+        {
+            error!(
+                "virtio queue used ring goes out of bounds: start:0x{:08x} size:0x{:08x}",
+                used_ring.raw_value(),
+                used_ring_size
+            );
+            false
+        } else if desc_table.mask(0xf) != 0 {
+            error!("virtio queue descriptor table breaks alignment contraints");
+            false
+        } else if avail_ring.mask(0x1) != 0 {
+            error!("virtio queue available ring breaks alignment contraints");
+            false
+        } else if used_ring.mask(0x3) != 0 {
+            error!("virtio queue used ring breaks alignment contraints");
+            false
+        } else {
+            true
+        }
+    }
+
+    /// A consuming iterator over all available descriptor chain heads offered by the driver.
+    pub fn iter<'a, 'b>(&'b mut self, mem: &'a GuestMemoryMmap) -> AvailIter<'a, 'b> {
+        let queue_size = self.actual_size();
+        let avail_ring = self.avail_ring;
+
+        let index_addr = match mem.checked_offset(avail_ring, 2) {
+            Some(ret) => ret,
+            None => {
+                // TODO log address
+                warn!("Invalid offset");
+                return AvailIter::new(mem, &mut self.next_avail);
+            }
+        };
+        // Note that last_index has no invalid values
+        let last_index: u16 = match mem.read_obj::<u16>(index_addr) {
+            Ok(ret) => ret,
+            Err(_) => return AvailIter::new(mem, &mut self.next_avail),
+        };
+
+        AvailIter {
+            mem,
+            desc_table: self.desc_table,
+            avail_ring,
+            next_index: self.next_avail,
+            last_index: Wrapping(last_index),
+            queue_size,
+            next_avail: &mut self.next_avail,
+        }
+    }
+
+    /// Puts an available descriptor head into the used ring for use by the guest.
+    pub fn add_used(&mut self, mem: &GuestMemoryMmap, desc_index: u16, len: u32) {
+        if desc_index >= self.actual_size() {
+            error!(
+                "attempted to add out of bounds descriptor to used ring: {}",
+                desc_index
+            );
+            return;
+        }
+
+        let used_ring = self.used_ring;
+        let next_used = (self.next_used.0 % self.actual_size()) as GuestUsize;
+        let used_elem = used_ring.unchecked_add(4 + next_used * 8);
+
+        // These writes can't fail as we are guaranteed to be within the descriptor ring.
+        mem.write_obj(u32::from(desc_index), used_elem).unwrap();
+        mem.write_obj(len as u32, used_elem.unchecked_add(4))
+            .unwrap();
+
+        self.next_used += Wrapping(1);
+
+        // This fence ensures all descriptor writes are visible before the index update is.
+        fence(Ordering::Release);
```

Comment:
This is not enough because `mem.write_obj` is not an atomic write. In practice it should work, but it would be better to use VolatileAtomicRef after rust-vmm/vm-memory#16 is fixed. Add a TODO?

Reply:
I will create an issue once this PR gets merged.